### PR TITLE
Support deploying via Github Actions using deploy --ci command

### PR DESCRIPTION
**Pull request by raysonkoh**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

Related to #1432 

**Overview of changes:**

Add logic for detecting Github Action CI environment as well as Repo Slug using environmental variables (`process.env.GITHUB_ACTIONS` and `process.env.GITHUB_REPOSITORY` respectively). 

**Anything you'd like to highlight / discuss:**
- Similar to deploying via Travis / Appveyor, the user has to generate a github personal access token with **repo** permissions and add a custom environmental variable. For Travis/Appveyor, the default name for such token in the userGuide is `GITHUB_TOKEN`. However, github [does not allow custom environmental variables with the `GITHUB_` prefix](https://docs.github.com/en/actions/reference/environment-variables#naming-conventions-for-environment-variables). Would it be better, for consistency, to update the default name for the token to just `TOKEN`? Or would it be better to create a special naming instruction for deploying via github actions?

**Testing instructions:**
Add to `.github/workflows/deploy.yml`:
```
name: Deploy
on:
  push:
    branches: master

jobs:
  build:
    runs-on: ubuntu-latest
    name: test
    env:
      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: '10'
      - run: cd ../ && git clone -b fix-1432-gh-action-2 https://github.com/raysonkoh/markbind.git
      - run: cd ../markbind/packages/cli && npm link
      - run: cd ../markbind && npm run setup
      - run: markbind build
      - run: markbind deploy --ci
```
For regression testing:
**Appveyor**: Add to `appveyor.yml`
```
environment:
  nodejs_version: '10'

branches:
  only:
    - master

install:
  - ps: Install-Product node $env:nodejs_version
  - cd ../
  - git clone -b fix-1432-gh-action-2 https://github.com/raysonkoh/markbind.git
  - cd markbind/packages/cli 
  - npm link 
  - cd ../../ 
  - npm run setup 
  - cd %APPVEYOR_BUILD_FOLDER%
  - markbind build
  - markbind deploy --ci

test: off

build: off
```
**Travis CI**: Add to `.travis.yml`
```
language: node_js
node_js:
  - 10
install:
  - cd ../ 
  - git clone -b fix-1432-gh-action-2 https://github.com/raysonkoh/markbind.git 
  - cd markbind/packages/cli 
  - npm link 
  - cd ../../ 
  - npm run setup 
  - cd $TRAVIS_BUILD_DIR
script: markbind build
deploy:
  provider: script
  script: markbind deploy --ci
  skip_cleanup: true
  on:
    branch: master
```

**Proposed commit message: (wrap lines at 72 characters)**
Support deploying via Github Actions using deploy --ci command

A Github Personal Access Token must be supplied as a
custom environmental variable in order to successfully deploy the
site via Travis CI and Appveyor.

For Github Actions, a default Github Token is
automatically generated for each Github Action job. The default Github
Token has sufficient write permissions to the repo, and so it
can be used in place of the Github Personal Access Token. To
incorporate the default Github Token, the deploy url has to use the
full url syntax of username:password. 

Let's update the deploy url to use x-access-token as the username, as
per the official Github documentation.

---

**Checklist:** :ballot_box_with_check:

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No blatantly unrelated changes
- [x] Pinged someone for a review!

### CODE REVIEW

**Comment by ang-zeyu** [comment]
> Similar to deploying via Travis / Appveyor, the user has to generate a github personal access token with repo permissions and add a custom environmental variable. For Travis/Appveyor, the default name for such token in the userGuide is GITHUB_TOKEN. However, github does not allow custom environmental variables with the GITHUB_ prefix. Would it be better, for consistency, to update the default name for the token to just TOKEN? Or would it be better to create a special naming instruction for deploying via github actions?

hmm looks like github actions actually has a `GITHUB_TOKEN` setup by default

https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token
are the permissions enough?

**Comment by ang-zeyu** [comment]
> Would it be better, for consistency, to update the default name for the token to just TOKEN? Or would it be better to create a special naming instruction for deploying via github actions?

Might be better to minimise the work needed to migrate over for now (special naming instruction), since the option itself was renamed.

---

On a unrelated note: this PR would make #1386 easier 🙂

**Comment by raysonkoh** [comment]
> https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token
are the permissions enough?

I tried using the default `GITHUB_TOKEN` but it received an error
```
 Unhandled rejection ProcessError: fatal: could not read Password for 'https://***@github.com': No such device or address

    at ChildProcess.child.on.code (/home/runner/work/test-markbind/markbind/node_modules/gh-pages/lib/git.js:42:16)
    at ChildProcess.emit (events.js:198:13)
    at maybeClose (internal/child_process.js:982:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
```
so I think it might not have the sufficient permissions. It works currently if I am using the generated personal access token.

**Comment by ang-zeyu** [comment]
From the docs:
> When you enable GitHub Actions, GitHub installs a GitHub App on your repository. The GITHUB_TOKEN secret is a **GitHub App installation access token.** You can use the installation access token to authenticate on behalf of the GitHub App installed on your repository. The token's permissions are limited to the repository that contains your workflow.

using the full url syntax with any username (`https://username:{GITHUB_TOKEN}/ .....`) instead seems to work (just my wild guess: likely due to above quote). Could you check if there are any security concerns with this?

If not, maybe we could auto-prepend something like `github-actions:{GITHUB_TOKEN}` for this special case.
Or we can leave it out, and require setting a secret pointing to a personal access token for standardisation.

**Comment by raysonkoh** [comment]
@ang-zeyu , I've checked that using `username:${GITHUB_TOKEN}` works. 

> Could you check if there are any security concerns (or anything else) with this?

There does not seem to be a lot of documentation regarding security of the default `GITHUB_TOKEN`. Based on the docs at https://docs.github.com/en/actions/reference/authentication-in-a-workflow#about-the-github_token-secret : 

> Before each job begins, GitHub fetches an installation access token for the job. The token expires when the job is finished.

since it only valid for the duration of the Github Actions job, there should not be any security concerns regarding this (that is within our control). Also, this is also in line with [Github encouraging the use of token-based authentication in place of password-based authentication](https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/) .

> If not, maybe we could auto-prepend something like github-actions:{GITHUB_TOKEN} for this special case.
Or we can leave it out, and require setting a secret pointing to a personal access token for standardisation.

I think both works. For the former, it may be more convenient for the user, since they do not have to specifically set up a personal access token and custom environmental variable. What do you think?

**Comment by ang-zeyu** [comment]
> since it only valid for the duration of the Github Actions job, there should not be any security concerns regarding this (that is within our control). Also, this is also in line with Github encouraging the use of token-based authentication in place of password-based authentication .

ok, thanks for investigating this. @raysonkoh 

I'm tending toward the former if possible since it requires zero setup, unless prepending an arbitary username wouldn't work with a personal / org access token in github actions (as there's no way currently to check what type the token is). 

In which case, we might need to introduce an additional user side option for identifying the github app tokenwhich dosen't seem worth it.

**Comment by ang-zeyu** [comment]
could get some others inputs too: @le0tan 

are you aware of any other concerns with prepending a username like so? `https://<any_username>:...token@github.com/owner/repo`

we're looking to integrate the token provided in github actions by default but it requires prepending `<any_username>:` (whereas personal / org access tokens don't)

**Comment by le0tan** [comment]
@ang-zeyu 

> could get some others inputs too: @le0tan
> 
> are you aware of any other concerns with prepending a username like so? `https://<any_username>:...token@github.com/owner/repo`
> 
> we're looking to integrate the token provided in github actions by default but it requires prepending `<any_username>:` (whereas personal / org access tokens don't)

This piece of documentation could be helpful:

https://github.com/MarkBind/markbind/pull/1460#issuecomment-770171351

in which for some reason unknown to me, they used `x-access-token` as the "username".

**Comment by ang-zeyu** [comment]
thanks for the info @le0tan 

@raysonkoh let's follow the username `x-access-token` as written in the github docs since there isn't much info on this.

> unless prepending an arbitary username wouldn't work with a personal / org access token in github actions (as there's no way currently to check what type the token is).

To elaborate a little, if we can't check internally what type of token is in the environment variable, we would likely have to introduce some sort of additional user configuration (e.g. `markbind deploy --gh-actions-app-token` or a necessitate a fixed environment variable like `GITHUB_APP_TOKEN`) which somewhat defeats the purpose of using the app token (little to no config knowledge required).

**Comment by raysonkoh** [comment]
Thanks for the quick response @ang-zeyu and @le0tan !

> To elaborate a little, if we can't check internally what type of token is in the environment variable, we would likely have to introduce some sort of additional user configuration (e.g. markbind deploy --gh-actions-app-token or a necessitate a fixed environment variable like GITHUB_APP_TOKEN) which somewhat defeats the purpose of using the app token (little to no config knowledge required).

Seems like prepending `x-access-token` still works with a personal access token. I have added deploy scripts for the other CI platforms in the PR description for your information. These are the cases that I have checked to be working on my personal test repo:
- Github Actions
  - Default `GITHUB_TOKEN`
  - Generated Personal Access Token
- Travis CI
  - Generated Personal Access Token
- Appveyor
  - Generated Personal Access Token

File: docs/userGuide/deployingTheSite.md
Comment (le0tan) [inline docs/userGuide/deployingTheSite.md:89]: is this a typo?

**Review by le0tan — CHANGES_REQUESTED** [review]

File: docs/userGuide/deployingTheSite.md
Comment (raysonkoh) [reply docs/userGuide/deployingTheSite.md]: Yes, thanks for catching it!

**Review by le0tan — COMMENTED** [review]
LGTM, pending @ang-zeyu 's approval

File: docs/userGuide/deployingTheSite.md
Comment (ang-zeyu) [inline docs/userGuide/deployingTheSite.md:97]: should we add a tip / note here that the example is using github's app token? 

Also that personal / org tokens would also work (can refer to travis section for setup instructions)

**Review by ang-zeyu — COMMENTED** [review]
Thanks! One more suggestion:

**Comment by ang-zeyu** [comment]
> Support deploying via Github Actions using deploy --ci command

let's update this to reflect why `x-access-token` was added in globally too

**Comment by raysonkoh** [comment]
@ang-zeyu I have updated the commit message, can you check whether it is okay?

File: docs/userGuide/deployingTheSite.md
Comment (raysonkoh) [reply docs/userGuide/deployingTheSite.md:102]: Yup I've added an info box for this.

**Review by ang-zeyu — APPROVED** [review]
Lgtm 👍 

> @ang-zeyu I have updated the commit message, can you check whether it is okay?

minor nit: we're following these [conventions](https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message) (link in pr template) for the commit message. Try to describe the current situation in present tense (e.g. no need for "previously"), then the action in imperative mood. (let's ... instead of was decided to...)

do wrap the message body at 72 characters too

**Comment by raysonkoh** [comment]
@ang-zeyu I have updated my commit message as per the convention. Will take note of this for future PRs!

**Pull request merged**
